rfcs/007-background-workers.md

```diff
@@ -0,0 +1,299 @@
+
+Workers can be dynamically started and stopped. When workers start, they MUST
+register themselves in the `"workers"` subspace based on their `WorkerType`.
+Each change to the workers list, should be accompanied by updating the
+`"workers_vs"` versionstamp. That versionstamp MAY be efficiently monitored for
+changes using a watch.
+
+
+### Worker Liveness Monitoring
+
+As long as the workers are active and ready to process jobs, they MUST
+periodically ping their health key. The period MUST be less than the
+`HealthTimeout` value.
+
+Worker nodes MUST monitor their neighbors for liveness. When they detect an
+unresponsive neighbor, they MUST remove their jobs and re-enqueue them in the
+pending queue, then remove the dead neighbor from the workers membership list
+as well. To avoid all worker nodes, monitoring all other workers, each worker
+will monitor their neighbor to "right" in the sorted list of live workers.
+
+### Liveness and Race Conditions
+
+In case workers come back after not updating their health entry often enough,
+and another worker had cleared and re-submitted their jobs, they MUST
+monitor their own presence in the membership list, and, if they detect they
+were removed, MUST stop running all their jobs. Then they MUST re-execute their
+startup logic: add themselves to the membership list, restart health key pings,
+and start picking jobs off of the pending jobs queue, and so on.
+
+### Clean Worker Shutdown
+
+If a worker is stopped cleanly, they MUST stop accepting new jobs from the
+pending queue, MUST stop running their jobs, MUST resubmit those jobs
+back into the pending queue. Finally, they MUST clear their health entry and
+remove themselves from workers membership list.
+
+### User Job Cancellation
+
+If a user decides to a cancel an active job, they MUST update the `CancelReq`
+field to `true`. Workers should periodically monitor that field and should stop
+running the job as soon as they detect it becoming `true`.
+
+## `/_active_tasks` API Compatibility
+
+The shape of the active table was deliberately constructed with the hopes of
+using it to implement the `/_active_tasks` API endpoint.
+
+When the HTTP API is accessed, a single `get_range(...)` request to
+`("couch_workers", "active")` subspace MAY be used to generate all the
+`_active_tasks` results. A job's `JobOpts` and `JobInfo` fields should contain
+most of needed information.
+
+# Advantages and Disadvantages
+
+Another alternative discussed on the mailing list is to mimic the
+pre-FoundationDB replication job scheduling model, where replication jobs are
+distributed between all the workers immediately when they are created. Each
+worker has their own separate job queue. The advantage of that model is that
+the queues and jobs are sharded evenly amongst the workers, avoiding potential
+hot key ranges around the global queue pushes and pops. The disadvantage is that it
+seemed like a premature optimization. Also, it would only allow for a completely
```

Comment:
This certain strategy prescribes premature, but sharded jobs could describe (additional job tuple term) parallelism opportunity, no reason workers must separately execute.  One worker can claim multiple jobs or ??

I don't see other sharding mentioned here, is other document concerning this with FDB?

Reply:
The sharding is happening in the current design (pre FDB) where each node is reading from it's own set of shards and running jobs from an in-memory queue (ets table).

> One worker can claim multiple jobs or ??

Right. The idea of worker is more of an application-type job manager. This manager would start multiple jobs of the same type, and then the would need to only maintain the liveness (health) of that supervisor or job manager. The equivalent would be something like the current replication scheduler.

However, after thinking about, it might be better to simplify and switch to a scheme with one worker being equivalent to a single process. So one worker executes on job only at a time...